### PR TITLE
Integrate Jörmungandr backend with CLI.

### PR DESCRIPTION
**Pull request by jonathanknowles**
# Issue Number

#357 

# Overview

This PR forks our wallet CLI into two separate executables:

1. **`cardano-wallet`**
    built with the `cardano-wallet-http-bridge` backend

2. **`cardano-wallet-jormungandr`**
    built with the `cardano-wallet-jormungandr` backend

Parts common to both executables are located within the `cardano-wallet-cli` library.

# Usage Examples

## Launching with HTTP bridge backend

```sh
stack exec cardano-wallet -- launch
```
## Launching with Jörmungandr backend

```sh
stack exec cardano-wallet-jormungandr -- launch \
    --backend-port=8081 \
    --genesis-hash=dba597bee5f0987efbf56f6bd7f44c38158a7770d0cb28a26b5eca40613a7ebd \
    --genesis-block=lib/jormungandr/test/data/jormungandr/block0.bin \
    --node-config=lib/jormungandr/test/data/jormungandr/config.yaml \
    --node-secret=lib/jormungandr/test/data/jormungandr/secret.yaml
```

# Work not tackled by this PR

There is a small amount of **code duplication** between both executables (the `execServe` functions in `exe/wallet/http-bridge/Main.hs` and `exe/wallet/jormungandr/Main.hs` are very similar). Ideally, we should reduce this duplication (probably suitable for a future PR).

### CODE REVIEW

File: exe/wallet/Main.hs [hunk 1]
```diff
@@ -411,7 +437,54 @@ execHttpBridge args _ = do
...
+newNetworkLayer
+    :: BaseUrl
+    -> Hash "Genesis"
+    -> IO (NetworkLayer (Jormungandr n) IO, Block Tx, FeePolicy)
+newNetworkLayer url block0H = do
+    mgr <- newManager defaultManagerSettings
+    let jormungandr = Jormungandr.mkJormungandrLayer mgr url
+    let nl = Jormungandr.mkNetworkLayer jormungandr
+    waitForConnection nl defaultRetryPolicy
+    block0 <- unsafeRunExceptT $ getBlock jormungandr (coerce block0H)
```
Comment (Anviking) [inline exe/wallet/Main.hs]: FYI, we *can* retrieve `block0H` ourselves

```haskell
{-# LANGUAGE OverloadedStrings #-}
import Cardano.Wallet.Jormungandr.Network
import Cardano.Wallet.Primitive.Types
import Cardano.Wallet.Unsafe
import Data.Coerce
    ( coerce )

main :: IO ()
main = do
    mgr <- newManager defaultManagerSettings
    let j = mkJormungandrLayer mgr (BaseUrl Http "localhost" 8081 "/api")
    let absentParent = Hash "\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL"
    [block0H] <- unsafeRunExceptT $ getDescendantIds j absentParent 1
    block0 <- unsafeRunExceptT $ getBlock j (coerce block0H)
    print block0
```

Comment (KtorZ) [reply exe/wallet/Main.hs]: Yes and no. We do want the user to provide it such that we can verify they match!

File: lib/cli/src/Cardano/CLI.hs [hunk 2]
```diff
@@ -130,14 +212,371 @@ import System.IO
...
+        <> cliExamples
+    cliCommands = unlines $ filter (not . null) $ lines $ mempty
+        <> cliCommandsSpecific
+        <> cliCommandsGeneric
+    cliOptions = unlines $ filter (not . null) $ sort $ lines $ mempty
+        <> cliOptionsSpecific
+        <> cliOptionsGeneric
+    cliExamples = unlines $ filter (not . null) $ lines $ mempty
+        <> cliExamplesSpecific
+        <> cliExamplesGeneric
```
Comment (KtorZ) [inline lib/cli/src/Cardano/CLI.hs:276]: :thinking: I think we're loosing the benefits of using `docopt` in a first place. I understand the need for removing duplication between both files yet I am not sure we are taking it to the right direction :/ .. It's probably the right time to switch to `optparse-applicative` for the CLI.

**Review by KtorZ — APPROVED** [review]

File: lib/cli/src/Cardano/CLI.hs (hunk 2 above)
Comment (jonathanknowles) [reply lib/cli/src/Cardano/CLI.hs:276]: >  I think we're loosing the benefits of using docopt in a first place. I understand the need for removing duplication between both files yet I am not sure we are taking it to the right direction :/ .. It's probably the right time to switch to optparse-applicative for the CLI.

I completely agree. Looking forward to trying a different direction with `optparse-applicative` here.

**Pull request merged**